python/cudf_polars/cudf_polars/experimental/rapidsmpf/shuffle.py

```diff
@@ -0,0 +1,269 @@
+    ir
+        The Shuffle IR node.
+    ch_in
+        Input ChannelPair with metadata and data channels.
+    ch_out
+        Output ChannelPair with metadata and data channels.
+    columns_to_hash
+        Tuple of column indices to use for hashing.
+    num_partitions
+        Number of partitions to shuffle into.
+    """
+    async with shutdown_on_error(
+        ctx, ch_in.metadata, ch_in.data, ch_out.metadata, ch_out.data
+    ):
+        # Create LocalShuffle context manager to handle shuffler lifecycle
+        with LocalShuffle(ctx, num_partitions, columns_to_hash) as local_shuffle:
+            # Process input chunks
+            while True:
+                msg = await ch_in.data.recv(ctx)
+                if msg is None:
+                    break
+
+                # Extract TableChunk from message
+                chunk = TableChunk.from_message(msg)
+
+                # Get the table view and insert into shuffler
+                table = chunk.table_view()
+                local_shuffle.insert_chunk(table)
+
+            # Extract shuffled partitions and send them out
+            # LocalShuffle.extract_chunk handles insert_finished, wait, extract, and unpack
+            for partition_id in range(num_partitions):
+                result_table = local_shuffle.extract_chunk(partition_id)
+
+                # Create a new TableChunk with the result
+                output_chunk = TableChunk.from_pylibcudf_table(
+                    sequence_number=partition_id,
+                    table=result_table,
+                    stream=local_shuffle.stream,
+                    exclusive_view=True,
+                )
+
+                # Send the output chunk
+                await ch_out.data.send(ctx, Message(output_chunk))
+
+        await ch_out.data.drain(ctx)
+
+
+@generate_ir_sub_network.register(Shuffle)
+def _(ir: Shuffle, rec: SubNetGenerator) -> tuple[list[Any], dict[IR, ChannelManager]]:
+    # Local shuffle operation.
+
+    # Process children
+    (child,) = ir.children
+    nodes, channels = rec(child)
+
+    keys: list[Col] = [ne.value for ne in ir.keys if isinstance(ne.value, Col)]
+    if len(keys) != len(ir.keys):  # pragma: no cover
+        raise NotImplementedError("Shuffle requires simple keys.")
+    column_names = list(ir.schema.keys())
+
+    context = rec.state["ctx"]
+    columns_to_hash = tuple(column_names.index(k.name) for k in keys)
+    num_partitions = rec.state["partition_info"][ir].count
+
+    # Create output ChannelManager
+    channels[ir] = ChannelManager()
+
+    # Complete shuffle pipeline in a single node
+    # LocalShuffle context manager handles shuffle ID lifecycle internally
+    nodes.append(
+        local_shuffle_node(
+            context,
+            ir,
+            ch_in=channels[child].reserve_output_slot(),
+            ch_out=channels[ir].reserve_input_slot(),
+            columns_to_hash=columns_to_hash,
+            num_partitions=num_partitions,
+        )
+    )
```

Comment:
Can you use the existing rapidsmpf facilities for this, rather than rolling your own?

Specifically:

- `rapidsmpf.streaming.cudf.partition.partition_and_pack`: packs input tablechunk channel into output partitionedmapchunks
- `rapidsmpf.streaming.coll.shuffler.shuffler`: accepts partitionedmapchunks and sends partitionedvectorchunks
- `rapidsmpf.streaming.cudf.partition.unpack_and_concat`: accepts partitionedvectorchunks and sends tablechunks